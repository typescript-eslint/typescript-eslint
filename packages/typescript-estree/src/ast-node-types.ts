/**
 * @fileoverview The AST node types produced by the parser.
 * @author Nicholas C. Zakas
 * @author James Henry <https://github.com/JamesHenry>
 * @copyright jQuery Foundation and other contributors, https://jquery.org/
 * MIT License
 */
export const AST_NODE_TYPES: { [key: string]: string } = {
  ArrayExpression: 'ArrayExpression',
  ArrayPattern: 'ArrayPattern',
  ArrowFunctionExpression: 'ArrowFunctionExpression',
  AssignmentExpression: 'AssignmentExpression',
  AssignmentPattern: 'AssignmentPattern',
  AwaitExpression: 'AwaitExpression',
  BigIntLiteral: 'BigIntLiteral',
  BinaryExpression: 'BinaryExpression',
  BlockStatement: 'BlockStatement',
  BreakStatement: 'BreakStatement',
  CallExpression: 'CallExpression',
  CatchClause: 'CatchClause',
  ClassBody: 'ClassBody',
  ClassDeclaration: 'ClassDeclaration',
  ClassExpression: 'ClassExpression',
  ClassImplements: 'ClassImplements',
  ClassProperty: 'ClassProperty',
  ConditionalExpression: 'ConditionalExpression',
  ContinueStatement: 'ContinueStatement',
  DebuggerStatement: 'DebuggerStatement',
  Decorator: 'Decorator',
  DoWhileStatement: 'DoWhileStatement',
  EmptyStatement: 'EmptyStatement',
  ExportAllDeclaration: 'ExportAllDeclaration',
  ExportDefaultDeclaration: 'ExportDefaultDeclaration',
  ExportNamedDeclaration: 'ExportNamedDeclaration',
  ExportSpecifier: 'ExportSpecifier',
  ExpressionStatement: 'ExpressionStatement',
  ForInStatement: 'ForInStatement',
  ForOfStatement: 'ForOfStatement',
  ForStatement: 'ForStatement',
  FunctionDeclaration: 'FunctionDeclaration',
  FunctionExpression: 'FunctionExpression',
  GenericTypeAnnotation: 'GenericTypeAnnotation',
  Identifier: 'Identifier',
  IfStatement: 'IfStatement',
  Import: 'Import',
  ImportDeclaration: 'ImportDeclaration',
  ImportDefaultSpecifier: 'ImportDefaultSpecifier',
  ImportNamespaceSpecifier: 'ImportNamespaceSpecifier',
  ImportSpecifier: 'ImportSpecifier',
  JSXAttribute: 'JSXAttribute',
  JSXClosingElement: 'JSXClosingElement',
  JSXClosingFragment: 'JSXClosingFragment',
  JSXElement: 'JSXElement',
  JSXEmptyExpression: 'JSXEmptyExpression',
  JSXExpressionContainer: 'JSXExpressionContainer',
  JSXFragment: 'JSXFragment',
  JSXIdentifier: 'JSXIdentifier',
  JSXMemberExpression: 'JSXMemberExpression',
  JSXNamespacedName: 'JSXNamespacedName',
  JSXOpeningElement: 'JSXOpeningElement',
  JSXOpeningFragment: 'JSXOpeningFragment',
  JSXSpreadAttribute: 'JSXSpreadAttribute',
  JSXSpreadChild: 'JSXSpreadChild',
  JSXText: 'JSXText',
  LabeledStatement: 'LabeledStatement',
  Literal: 'Literal',
  LogicalExpression: 'LogicalExpression',
  MemberExpression: 'MemberExpression',
  MetaProperty: 'MetaProperty',
  MethodDefinition: 'MethodDefinition',
  NewExpression: 'NewExpression',
  ObjectExpression: 'ObjectExpression',
  ObjectPattern: 'ObjectPattern',
  Program: 'Program',
  Property: 'Property',
  RestElement: 'RestElement',
  ReturnStatement: 'ReturnStatement',
  SequenceExpression: 'SequenceExpression',
  SpreadElement: 'SpreadElement',
  Super: 'Super',
  SwitchCase: 'SwitchCase',
  SwitchStatement: 'SwitchStatement',
  TaggedTemplateExpression: 'TaggedTemplateExpression',
  TemplateElement: 'TemplateElement',
  TemplateLiteral: 'TemplateLiteral',
  ThisExpression: 'ThisExpression',
  ThrowStatement: 'ThrowStatement',
  TryStatement: 'TryStatement',
  UnaryExpression: 'UnaryExpression',
  UpdateExpression: 'UpdateExpression',
  VariableDeclaration: 'VariableDeclaration',
  VariableDeclarator: 'VariableDeclarator',
  WhileStatement: 'WhileStatement',
  WithStatement: 'WithStatement',
  YieldExpression: 'YieldExpression',
  /**
   * TS-prefixed nodes
   */
  TSAbstractClassDeclaration: 'TSAbstractClassDeclaration',
  TSAbstractClassProperty: 'TSAbstractClassProperty',
  TSAbstractKeyword: 'TSAbstractKeyword',
  TSAbstractMethodDefinition: 'TSAbstractMethodDefinition',
  TSAnyKeyword: 'TSAnyKeyword',
  TSArrayType: 'TSArrayType',
  TSAsExpression: 'TSAsExpression',
  TSAsyncKeyword: 'TSAsyncKeyword',
  TSBooleanKeyword: 'TSBooleanKeyword',
  TSBigIntKeyword: 'TSBigIntKeyword',
  TSConditionalType: 'TSConditionalType',
  TSConstructorType: 'TSConstructorType',
  TSConstructSignature: 'TSConstructSignature',
  TSDeclareKeyword: 'TSDeclareKeyword',
  TSDeclareFunction: 'TSDeclareFunction',
  TSEnumDeclaration: 'TSEnumDeclaration',
  TSEnumMember: 'TSEnumMember',
  TSExportAssignment: 'TSExportAssignment',
  TSExportKeyword: 'TSExportKeyword',
  TSImportType: 'TSImportType',
  TSInferType: 'TSInferType',
  TSLiteralType: 'TSLiteralType',
  TSIndexedAccessType: 'TSIndexedAccessType',
  TSIndexSignature: 'TSIndexSignature',
  TSInterfaceBody: 'TSInterfaceBody',
  TSInterfaceDeclaration: 'TSInterfaceDeclaration',
  TSInterfaceHeritage: 'TSInterfaceHeritage',
  TSFunctionType: 'TSFunctionType',
  TSMethodSignature: 'TSMethodSignature',
  TSModuleBlock: 'TSModuleBlock',
  TSModuleDeclaration: 'TSModuleDeclaration',
  TSNamespaceFunctionDeclaration: 'TSNamespaceFunctionDeclaration',
  TSNonNullExpression: 'TSNonNullExpression',
  TSNeverKeyword: 'TSNeverKeyword',
  TSNullKeyword: 'TSNullKeyword',
  TSNumberKeyword: 'TSNumberKeyword',
  TSMappedType: 'TSMappedType',
  TSObjectKeyword: 'TSObjectKeyword',
  TSParameterProperty: 'TSParameterProperty',
  TSPrivateKeyword: 'TSPrivateKeyword',
  TSPropertySignature: 'TSPropertySignature',
  TSProtectedKeyword: 'TSProtectedKeyword',
  TSPublicKeyword: 'TSPublicKeyword',
  TSQualifiedName: 'TSQualifiedName',
  TSQuestionToken: 'TSQuestionToken',
  TSReadonlyKeyword: 'TSReadonlyKeyword',
  TSRestType: 'TSRestType',
  TSStaticKeyword: 'TSStaticKeyword',
  TSStringKeyword: 'TSStringKeyword',
  TSSymbolKeyword: 'TSSymbolKeyword',
  TSThisType: 'TSThisType',
  TSTypeAnnotation: 'TSTypeAnnotation',
  TSTypeAliasDeclaration: 'TSTypeAliasDeclaration',
  TSTypeLiteral: 'TSTypeLiteral',
  TSTypeOperator: 'TSTypeOperator',
  TSTypeParameter: 'TSTypeParameter',
  TSTypeParameterDeclaration: 'TSTypeParameterDeclaration',
  TSTypeParameterInstantiation: 'TSTypeParameterInstantiation',
  TSTypePredicate: 'TSTypePredicate',
  TSTypeReference: 'TSTypeReference',
  TSTypeQuery: 'TSTypeQuery',
  TSIntersectionType: 'TSIntersectionType',
  TSTupleType: 'TSTupleType',
  TSOptionalType: 'TSOptionalType',
  TSParenthesizedType: 'TSParenthesizedType',
  TSUnionType: 'TSUnionType',
  TSUndefinedKeyword: 'TSUndefinedKeyword',
  TSUnknownKeyword: 'TSUnknownKeyword',
  TSVoidKeyword: 'TSVoidKeyword'
};
