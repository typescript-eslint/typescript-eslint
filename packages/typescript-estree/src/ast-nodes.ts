import * as es from './ast-tree-nodes';

export type ASTTreeNodes =
  | es.ArrayExpression
  | es.ArrayPattern
  | es.ArrowFunctionExpression
  | es.AssignmentExpression
  | es.AssignmentPattern
  | es.AwaitExpression
  | es.BigIntLiteral
  | es.BinaryExpression
  | es.BlockStatement
  | es.BreakStatement
  | es.CallExpression
  | es.CatchClause
  | es.ClassBody
  | es.ClassDeclaration
  | es.ClassExpression
  | es.ClassProperty
  | es.ConditionalExpression
  | es.ContinueStatement
  | es.DebuggerStatement
  | es.Decorator
  | es.DoWhileStatement
  | es.EmptyStatement
  | es.ExportAllDeclaration
  | es.ExportDefaultDeclaration
  | es.ExportNamedDeclaration
  | es.ExportSpecifier
  | es.ExpressionStatement
  | es.ForInStatement
  | es.ForOfStatement
  | es.ForStatement
  | es.FunctionDeclaration
  | es.FunctionExpression
  | es.Identifier
  | es.IfStatement
  | es.Import
  | es.ImportDeclaration
  | es.ImportDefaultSpecifier
  | es.ImportNamespaceSpecifier
  | es.ImportSpecifier
  | es.JSXAttribute
  | es.JSXClosingElement
  | es.JSXClosingFragment
  | es.JSXElement
  | es.JSXEmptyExpression
  | es.JSXExpressionContainer
  | es.JSXFragment
  | es.JSXIdentifier
  | es.JSXMemberExpression
  // | es.JSXNamespacedName // https://github.com/Microsoft/TypeScript/issues/7411
  | es.JSXOpeningElement
  | es.JSXOpeningFragment
  | es.JSXSpreadAttribute
  | es.JSXSpreadChild
  | es.JSXText
  | es.LabeledStatement
  | es.Literal
  | es.LogicalExpression
  | es.MemberExpression
  | es.MetaProperty
  | es.MethodDefinition
  | es.NewExpression
  | es.ObjectExpression
  | es.ObjectPattern
  | es.Program
  | es.Property
  | es.RestElement
  | es.ReturnStatement
  | es.SequenceExpression
  | es.SpreadElement
  | es.Super
  | es.SwitchCase
  | es.SwitchStatement
  | es.TaggedTemplateExpression
  | es.TemplateElement
  | es.TemplateLiteral
  | es.ThisExpression
  | es.ThrowStatement
  | es.TryStatement
  | es.UnaryExpression
  | es.UpdateExpression
  | es.VariableDeclaration
  | es.VariableDeclarator
  | es.WhileStatement
  | es.WithStatement
  | es.YieldExpression
  | es.TSAbstractClassProperty
  // | es.TSAbstractKeyword
  | es.TSAbstractMethodDefinition
  | es.TSAnyKeyword
  | es.TSArrayType
  | es.TSAsExpression
  | es.TSAsyncKeyword
  | es.TSBooleanKeyword
  | es.TSBigIntKeyword
  | es.TSConditionalType
  | es.TSConstructorType
  | es.TSCallSignatureDeclaration
  | es.TSClassImplements
  | es.TSConstructSignatureDeclaration
  // | es.TSDeclareKeyword
  | es.TSDeclareFunction
  | es.TSEnumDeclaration
  | es.TSEnumMember
  | es.TSExportAssignment
  // | es.TSExportKeyword
  | es.TSExternalModuleReference
  | es.TSImportType
  | es.TSInferType
  | es.TSLiteralType
  | es.TSIndexedAccessType
  | es.TSIndexSignature
  | es.TSInterfaceBody
  | es.TSInterfaceDeclaration
  | es.TSInterfaceHeritage
  | es.TSImportEqualsDeclaration
  | es.TSFunctionType
  | es.TSMethodSignature
  | es.TSModuleBlock
  | es.TSModuleDeclaration
  | es.TSNamespaceExportDeclaration
  | es.TSNonNullExpression
  | es.TSNeverKeyword
  | es.TSNullKeyword
  | es.TSNumberKeyword
  | es.TSMappedType
  | es.TSObjectKeyword
  | es.TSParameterProperty
  | es.TSPrivateKeyword
  | es.TSPropertySignature
  | es.TSProtectedKeyword
  | es.TSPublicKeyword
  | es.TSQualifiedName
  // | es.TSQuestionToken
  // | es.TSReadonlyKeyword
  | es.TSRestType
  | es.TSStaticKeyword
  | es.TSStringKeyword
  | es.TSSymbolKeyword
  | es.TSThisType
  | es.TSTypeAnnotation
  | es.TSTypeAliasDeclaration
  | es.TSTypeAssertion
  | es.TSTypeLiteral
  | es.TSTypeOperator
  | es.TSTypeParameter
  | es.TSTypeParameterDeclaration
  | es.TSTypeParameterInstantiation
  | es.TSTypePredicate
  | es.TSTypeReference
  | es.TSTypeQuery
  | es.TSIntersectionType
  | es.TSTupleType
  | es.TSOptionalType
  | es.TSParenthesizedType
  | es.TSUnionType
  | es.TSUndefinedKeyword
  | es.TSUnknownKeyword
  | es.TSVoidKeyword
  // abstract types
  | es.FunctionDeclarationLike
  | es.ClassPropertyLikeNode
  | es.ClassLikeNode
  | es.TSFunctionDeclarationLike
  | es.TSExpressionWithTypeArgumentsLike;
