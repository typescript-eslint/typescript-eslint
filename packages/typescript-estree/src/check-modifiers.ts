import * as ts from 'typescript';

import type { TSNode } from './ts-estree';

import { getDecorators, getModifiers } from './getModifiers';
import {
  getDeclarationKind,
  hasModifier,
  isThisIdentifier,
  createError,
} from './node-utils';

const SyntaxKind = ts.SyntaxKind;

// `ts.nodeIsMissing`
function nodeIsMissing(node: ts.Node | undefined): boolean {
  if (node == null) {
    return true;
  }
  return (
    node.pos === node.end &&
    node.pos >= 0 &&
    node.kind !== SyntaxKind.EndOfFileToken
  );
}

// `ts.nodeIsPresent`
function nodeIsPresent(node: ts.Node | undefined): node is ts.Node {
  return !nodeIsMissing(node);
}

// `ts.hasAbstractModifier`
function hasAbstractModifier(node: ts.Node): boolean {
  return hasModifier(SyntaxKind.AbstractKeyword, node);
}

// `ts.getThisParameter`
function getThisParameter(
  signature: ts.SignatureDeclaration,
): ts.ParameterDeclaration | null {
  if (signature.parameters.length && !ts.isJSDocSignature(signature)) {
    const thisParameter = signature.parameters[0];
    if (parameterIsThisKeyword(thisParameter)) {
      return thisParameter;
    }
  }

  return null;
}

// `ts.parameterIsThisKeyword`
function parameterIsThisKeyword(parameter: ts.ParameterDeclaration): boolean {
  return isThisIdentifier(parameter.name);
}

// `ts.getContainingFunction`
function getContainingFunction(
  node: ts.Node,
): ts.SignatureDeclaration | undefined {
  return ts.findAncestor(node.parent, ts.isFunctionLike);
}

// Rewrite version of `ts.nodeCanBeDecorated`
// Returns `true` for both `useLegacyDecorators: true` and `useLegacyDecorators: false`
function nodeCanBeDecorated(node: TSNode): boolean {
  switch (node.kind) {
    case SyntaxKind.ClassDeclaration:
      return true;
    case SyntaxKind.ClassExpression:
      // `ts.nodeCanBeDecorated` returns `false` if `useLegacyDecorators: true`
      return true;
    case SyntaxKind.PropertyDeclaration: {
      const { parent } = node;

      // `ts.nodeCanBeDecorated` uses this if `useLegacyDecorators: true`
      if (ts.isClassDeclaration(parent)) {
        return true;
      }

      // `ts.nodeCanBeDecorated` uses this if `useLegacyDecorators: false`
      if (ts.isClassLike(parent) && !hasAbstractModifier(node)) {
        return true;
      }

      return false;
    }
    case SyntaxKind.GetAccessor:
    case SyntaxKind.SetAccessor:
    case SyntaxKind.MethodDeclaration: {
      const { parent } = node;
      // In `ts.nodeCanBeDecorated`
      // when `useLegacyDecorators: true` uses `ts.isClassDeclaration`
      // when `useLegacyDecorators: true` uses `ts.isClassLike`
      return (
        Boolean(node.body) &&
        (ts.isClassDeclaration(parent) || ts.isClassLike(parent))
      );
    }
    case SyntaxKind.Parameter: {
      // `ts.nodeCanBeDecorated` returns `false` if `useLegacyDecorators: false`

      const { parent } = node;
      const grandparent = parent.parent;

      return (
        Boolean(parent) &&
        'body' in parent &&
        Boolean(parent.body) &&
        (parent.kind === SyntaxKind.Constructor ||
          parent.kind === SyntaxKind.MethodDeclaration ||
          parent.kind === SyntaxKind.SetAccessor) &&
        getThisParameter(parent) !== node &&
        Boolean(grandparent) &&
        grandparent.kind === SyntaxKind.ClassDeclaration
      );
    }
  }

  return false;
}

function nodeHasIllegalDecorators(
  node: ts.Node,
): node is { illegalDecorators: ts.Node[] } & ts.Node {
  return !!(
    'illegalDecorators' in node &&
    (node.illegalDecorators as unknown[] | undefined)?.length
  );
}

function throwError(node: ts.Node, message: string): never {
  const ast = node.getSourceFile();
  const start = node.getStart(ast);
  const end = node.getEnd();

  throw createError(message, ast, start, end);
}

export function checkModifiers(node: ts.Node): void {
  // typescript<5.0.0
  if (nodeHasIllegalDecorators(node)) {
    throwError(node.illegalDecorators[0], 'Decorators are not valid here.');
  }

  for (const decorator of getDecorators(
    node,
    /* includeIllegalDecorators */ true,
  ) ?? []) {
    // `checkGrammarModifiers` function in typescript
    if (!nodeCanBeDecorated(node as TSNode)) {
      if (ts.isMethodDeclaration(node) && !nodeIsPresent(node.body)) {
        throwError(
          decorator,
          'A decorator can only decorate a method implementation, not an overload.',
        );
      } else {
        throwError(decorator, 'Decorators are not valid here.');
      }
    }
  }

  for (const modifier of getModifiers(
    node,
    /* includeIllegalModifiers */ true,
  ) ?? []) {
    if (modifier.kind !== SyntaxKind.ReadonlyKeyword) {
      if (
        node.kind === SyntaxKind.PropertySignature ||
        node.kind === SyntaxKind.MethodSignature
      ) {
        throwError(
          modifier,
          `'${ts.tokenToString(
            modifier.kind,
          )}' modifier cannot appear on a type member`,
        );
      }

      if (
        node.kind === SyntaxKind.IndexSignature &&
        (modifier.kind !== SyntaxKind.StaticKeyword ||
          !ts.isClassLike(node.parent))
      ) {
        throwError(
          modifier,
          `'${ts.tokenToString(
            modifier.kind,
          )}' modifier cannot appear on an index signature`,
        );
      }
    }

    if (
      modifier.kind !== SyntaxKind.InKeyword &&
      modifier.kind !== SyntaxKind.OutKeyword &&
      modifier.kind !== SyntaxKind.ConstKeyword &&
      node.kind === SyntaxKind.TypeParameter
    ) {
      throwError(
        modifier,
        `'${ts.tokenToString(
          modifier.kind,
        )}' modifier cannot appear on a type parameter`,
      );
    }

    if (
      (modifier.kind === SyntaxKind.InKeyword ||
        modifier.kind === SyntaxKind.OutKeyword) &&
      (node.kind !== SyntaxKind.TypeParameter ||
        !(
          ts.isInterfaceDeclaration(node.parent) ||
          ts.isClassLike(node.parent) ||
          ts.isTypeAliasDeclaration(node.parent)
        ))
    ) {
      throwError(
        modifier,
        `'${ts.tokenToString(
          modifier.kind,
        )}' modifier can only appear on a type parameter of a class, interface or type alias`,
      );
    }

    if (
      modifier.kind === SyntaxKind.ReadonlyKeyword &&
      node.kind !== SyntaxKind.PropertyDeclaration &&
      node.kind !== SyntaxKind.PropertySignature &&
      node.kind !== SyntaxKind.IndexSignature &&
      node.kind !== SyntaxKind.Parameter
    ) {
      throwError(
        modifier,
        "'readonly' modifier can only appear on a property declaration or index signature.",
      );
    }

    if (
      modifier.kind === SyntaxKind.DeclareKeyword &&
      ts.isClassLike(node.parent) &&
      !ts.isPropertyDeclaration(node)
    ) {
      throwError(
        modifier,
        `'${ts.tokenToString(
          modifier.kind,
        )}' modifier cannot appear on class elements of this kind.`,
      );
    }

    if (
      modifier.kind === SyntaxKind.DeclareKeyword &&
      ts.isVariableStatement(node)
    ) {
      const declarationKind = getDeclarationKind(node.declarationList);
      if (declarationKind === 'using' || declarationKind === 'await using') {
        throwError(
          modifier,
          `'declare' modifier cannot appear on a '${declarationKind}' declaration.`,
        );
      }
    }

    if (
      modifier.kind === SyntaxKind.AbstractKeyword &&
      node.kind !== SyntaxKind.ClassDeclaration &&
      node.kind !== SyntaxKind.ConstructorType &&
      node.kind !== SyntaxKind.MethodDeclaration &&
      node.kind !== SyntaxKind.PropertyDeclaration &&
      node.kind !== SyntaxKind.GetAccessor &&
      node.kind !== SyntaxKind.SetAccessor
    ) {
      throwError(
        modifier,
        `'${ts.tokenToString(
          modifier.kind,
        )}' modifier can only appear on a class, method, or property declaration.`,
      );
    }

    if (
      (modifier.kind === SyntaxKind.StaticKeyword ||
        modifier.kind === SyntaxKind.PublicKeyword ||
        modifier.kind === SyntaxKind.ProtectedKeyword ||
        modifier.kind === SyntaxKind.PrivateKeyword) &&
      (node.parent.kind === SyntaxKind.ModuleBlock ||
        node.parent.kind === SyntaxKind.SourceFile)
    ) {
      throwError(
        modifier,
        `'${ts.tokenToString(
          modifier.kind,
        )}' modifier cannot appear on a module or namespace element.`,
      );
    }

    if (
      modifier.kind === SyntaxKind.AccessorKeyword &&
      node.kind !== SyntaxKind.PropertyDeclaration
    ) {
      throwError(
        modifier,
        "'accessor' modifier can only appear on a property declaration.",
      );
    }

    // `checkGrammarAsyncModifier` function in `typescript`
    if (
      modifier.kind === SyntaxKind.AsyncKeyword &&
      node.kind !== SyntaxKind.MethodDeclaration &&
      node.kind !== SyntaxKind.FunctionDeclaration &&
      node.kind !== SyntaxKind.FunctionExpression &&
      node.kind !== SyntaxKind.ArrowFunction
    ) {
      throwError(modifier, "'async' modifier cannot be used here.");
    }

    // `checkGrammarModifiers` function in `typescript`
    if (
      node.kind === SyntaxKind.Parameter &&
      (modifier.kind === SyntaxKind.StaticKeyword ||
        modifier.kind === SyntaxKind.ExportKeyword ||
        modifier.kind === SyntaxKind.DeclareKeyword ||
        modifier.kind === SyntaxKind.AsyncKeyword)
    ) {
      throwError(
        modifier,
        `'${ts.tokenToString(
          modifier.kind,
        )}' modifier cannot appear on a parameter.`,
      );
    }

    // `checkGrammarModifiers` function in `typescript`
    if (
      modifier.kind === SyntaxKind.PublicKeyword ||
      modifier.kind === SyntaxKind.ProtectedKeyword ||
      modifier.kind === SyntaxKind.PrivateKeyword
    ) {
      for (const anotherModifier of getModifiers(node) ?? []) {
        if (
          anotherModifier !== modifier &&
          (anotherModifier.kind === SyntaxKind.PublicKeyword ||
            anotherModifier.kind === SyntaxKind.ProtectedKeyword ||
            anotherModifier.kind === SyntaxKind.PrivateKeyword)
        ) {
          throwError(anotherModifier, `Accessibility modifier already seen.`);
        }
      }
    }

    // `checkParameter` function in `typescript`
    if (
      node.kind === SyntaxKind.Parameter &&
      // In `typescript` package, it's `ts.hasSyntacticModifier(node, ts.ModifierFlags.ParameterPropertyModifier)`
      // https://github.com/typescript-eslint/typescript-eslint/pull/6615#discussion_r1136489935
      (modifier.kind === SyntaxKind.PublicKeyword ||
        modifier.kind === SyntaxKind.PrivateKeyword ||
        modifier.kind === SyntaxKind.ProtectedKeyword ||
        modifier.kind === SyntaxKind.ReadonlyKeyword ||
        modifier.kind === SyntaxKind.OverrideKeyword)
    ) {
      const func = getContainingFunction(node);

      if (
        !(func?.kind === SyntaxKind.Constructor && nodeIsPresent(func.body))
      ) {
        throwError(
          modifier,
          'A parameter property is only allowed in a constructor implementation.',
        );
      }
    }

    // There are more cases in `checkGrammarObjectLiteralExpression` in TypeScript.
    // We may add more validations for them here in the future.
    if (
      (modifier.kind !== SyntaxKind.AsyncKeyword ||
        node.kind !== SyntaxKind.MethodDeclaration) &&
      node.parent.kind === SyntaxKind.ObjectLiteralExpression
    ) {
      throwError(
        modifier,
        `'${ts.tokenToString(modifier.kind)}' modifier cannot be used here.`,
      );
    }
  }
}
