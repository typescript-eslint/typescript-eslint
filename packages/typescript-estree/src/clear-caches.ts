import { clearWatchCaches } from './create-program/getWatchProgramsForProjects';
import { clearProgramCache as clearProgramCacheOriginal } from './parser';
import { clearTSConfigMatchCache } from './parseSettings/createParseSettings';
import { clearGlobCache } from './parseSettings/resolveProjectList';

/**
 * Clears all of the internal caches.
 * Generally you shouldn't need or want to use this.
 * Examples of intended uses:
 * - In tests to reset parser state to keep tests isolated.
 * - In custom lint tooling that iteratively lints one project at a time to prevent OOMs.
 */
export function clearCaches(): void {
  clearProgramCacheOriginal();
  clearWatchCaches();
  clearTSConfigMatchCache();
  clearGlobCache();
}

// TODO - delete this in next major
export const clearProgramCache = clearCaches();
