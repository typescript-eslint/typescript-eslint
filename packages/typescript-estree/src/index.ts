export * from './parser';
export { ParserServices, TSESTreeOptions } from './parser-options';
export { simpleTraverse } from './simple-traverse';
export * from './ts-estree';
export { clearCaches } from './create-program/createWatchProgram';

// re-export for backwards-compat
export { visitorKeys } from '@typescript-eslint/visitor-keys';

// note - cannot migrate this to an import statement because it will make TSC copy the package.json to the dist folder
export const version: string = require('../package.json').version;
