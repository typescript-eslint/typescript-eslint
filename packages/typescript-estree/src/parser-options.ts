import { Program } from 'typescript';
import { Token, Comment, Node } from './ts-estree/ts-estree';
import { TSNode } from './ts-nodes';

export interface Extra {
  errorOnUnknownASTType: boolean;
  errorOnTypeScriptSyntacticAndSemanticIssues: boolean;
  useJSXTextNode: boolean;
  tokens: null | Token[];
  comment: boolean;
  code: string;
  range: boolean;
  loc: boolean;
  comments: Comment[];
  strict: boolean;
  jsx: boolean;
  log: Function;
  projects: string[];
  tsconfigRootDir: string;
  extraFileExtensions: string[];
}

export interface ParserOptions {
  range?: boolean;
  loc?: boolean;
  tokens?: boolean;
  comment?: boolean;
  jsx?: boolean;
  errorOnUnknownASTType?: boolean;
  errorOnTypeScriptSyntacticAndSemanticIssues?: boolean;
  useJSXTextNode?: boolean;
  loggerFn?: Function | false;
  project?: string | string[];
  filePath?: string;
  tsconfigRootDir?: string;
  extraFileExtensions?: string[];
}

export interface ParserWeakMap<TKey, TValueBase> {
  get<TValue extends TValueBase>(key: TKey): TValue;
  has(key: any): boolean;
}

export interface ParserServices {
  program: Program | undefined;
  esTreeNodeToTSNodeMap: ParserWeakMap<Node, TSNode> | undefined;
  tsNodeToESTreeNodeMap: ParserWeakMap<TSNode, Node> | undefined;
}
