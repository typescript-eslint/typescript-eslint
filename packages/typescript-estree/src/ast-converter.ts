import { SourceFile } from 'typescript';
import { convertError, Converter, ASTMaps } from './convert';
import { convertComments } from './convert-comments';
import { convertTokens } from './node-utils';
import { Extra } from './parser-options';
import { TSESTree } from './ts-estree';
import { simpleTraverse } from './simple-traverse';

export function astConverter(
  ast: SourceFile,
  extra: Extra,
  shouldPreserveNodeMaps: boolean,
): { estree: TSESTree.Program; astMaps: ASTMaps | undefined } {
  /**
   * The TypeScript compiler produced fundamental parse errors when parsing the
   * source.
   */
  // internal typescript api...
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const parseDiagnostics = (ast as any).parseDiagnostics;
  if (parseDiagnostics.length) {
    throw convertError(parseDiagnostics[0]);
  }

  /**
   * Recursively convert the TypeScript AST into an ESTree-compatible AST
   */
  const instance = new Converter(ast, {
    errorOnUnknownASTType: extra.errorOnUnknownASTType || false,
    useJSXTextNode: extra.useJSXTextNode || false,
    shouldPreserveNodeMaps,
  });

  const estree = instance.convertProgram();

  /**
   * Optionally remove range and loc if specified
   */
  if (!extra.range || !extra.loc) {
    simpleTraverse(estree, {
      enter: node => {
        if (!extra.range) {
          delete node.range;
        }
        if (!extra.loc) {
          delete node.loc;
        }
      },
    });
  }

  /**
   * Optionally convert and include all tokens in the AST
   */
  if (extra.tokens) {
    estree.tokens = convertTokens(ast);
  }

  /**
   * Optionally convert and include all comments in the AST
   */
  if (extra.comment) {
    estree.comments = convertComments(ast, extra.code);
  }

  const astMaps = shouldPreserveNodeMaps ? instance.getASTMaps() : undefined;

  return { estree, astMaps };
}
