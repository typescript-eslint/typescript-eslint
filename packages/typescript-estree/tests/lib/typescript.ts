/**
 * @fileoverview Tests for TypeScript-specific constructs
 * @author Nicholas C. Zakas
 * @author James Henry <https://github.com/JamesHenry>
 * @copyright jQuery Foundation and other contributors, https://jquery.org/
 * MIT License
 */
import { readFileSync } from 'fs';
import glob from 'glob';
import { extname } from 'path';
import { ParserOptions } from '../../src/parser-options';
import {
  createSnapshotTestBlock,
  formatSnapshotName,
  isJSXFileType
} from '../../tools/test-utils';

//------------------------------------------------------------------------------
// Setup
//------------------------------------------------------------------------------

const FIXTURES_DIR =
  '../../node_modules/@typescript-eslint/shared-fixtures/fixtures/typescript';
const testFiles = glob.sync(`${FIXTURES_DIR}/**/*.src.ts`);

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

describe('typescript', () => {
  testFiles.forEach(filename => {
    const code = readFileSync(filename, 'utf8');
    const fileExtension = extname(filename);
    const config: ParserOptions = {
      loc: true,
      range: true,
      tokens: true,
      errorOnUnknownASTType: true,
      jsx: isJSXFileType(fileExtension)
    };
    it(
      formatSnapshotName(filename, FIXTURES_DIR, fileExtension),
      createSnapshotTestBlock(code, config)
    );
  });
});
