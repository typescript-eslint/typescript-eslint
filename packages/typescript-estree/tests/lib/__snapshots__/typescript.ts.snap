// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typescript fixtures/errorRecovery/interface-with-no-body.src 1`] = `"'{' expected."`;
